cmake/modules/vif.cmake

```diff
@@ -0,0 +1,33 @@
+# Copyright (c) 2022 The Chromium OS Authors
+# SPDX-License-Identifier: Apache-2.0
+
+if (CONFIG_GENERATE_VIF)
+    # Generates USB-C VIF policies in XML format from device tree.
+    set(GEN_VIF_SCRIPT ${ZEPHYR_BASE}/scripts/generate_usb_vif/generate_vif.py)
+    set(DTS_COMPATIBLE usb-c-connector)
+    set(VIF_XML ${PROJECT_BINARY_DIR}/vif.xml)
+    set(CMD_GEN_VIF ${PYTHON_EXECUTABLE} ${GEN_VIF_SCRIPT}
+            --edt-pickle ${EDT_PICKLE}
+            --compatible ${DTS_COMPATIBLE}
+            --vif-out ${VIF_XML}.new
+            --board ${BOARD}
+            )
+
+    message(STATUS "Generating vif.xml")
+
+    # Executes generate_vif.py to create vif.xml.
+    execute_process(
+            COMMAND ${CMD_GEN_VIF}
+            WORKING_DIRECTORY ${PROJECT_BINARY_DIR}
+            RESULT_VARIABLE ret
+    )
+
+    if (NOT "${ret}" STREQUAL "0")
+        message(STATUS "In: ${PROJECT_BINARY_DIR}, command: ${CMD_GEN_VIF}")
+        message(FATAL_ERROR "generate_vif.py failed with return code: ${ret}")
+    else ()
+        zephyr_file_copy(${VIF_XML}.new ${VIF_XML} ONLY_IF_DIFFERENT)
```

Comment:
Why does this need to generate the file every time it runs and then compare it to the already generated file, can you not check if the file needs regenerating because e.g. some input files have changed and then only regenerate if it needs to be regenerated?

Reply:
I kind of picked that mechanism from dts.cmake where zephyr.dts is being generated similar way. I wasn't sure how to compare if input files are changed. The DT file locations are unknown as they could be present in any project to check at DT level. At python script, I guess comparing vif.xml with past vif.xml is the only way as zephyr.dts.pre will be updated by that time with current DT version. If there are any other possible ways, could you please share them.

Reply:
This should be replaced with `add_custom_command`/`add_custom_target`, that way this can have proper dependency checking and will only run if cmake determines that it needs to be re-generated. You can see an example of it being used in-tree at https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/tfm_integration/psa_firmware/CMakeLists.txt#L24

Reply:
Do you mean replace zephyr_file_copy method with add_custom_method

Reply:
No, the execute command part, the process should only run to generate this output file if it needs to. There would then not be an intermediary output file, it would just create the file or overwrite the existing file

Reply:
"the process should only run to generate this output file if it needs to", here do you mean adding any checks before entering script, like if there is any difference w.r.to past version

Reply:
Not sure where your message is but no, given the same input files the output should be the same, cmake will handle this for you. If the input files change, then the output should be re-generated automatically.

Reply:
I deleted it as I understood it bit later, thanks, I added add_custom_command in .cmake file when I run script it doesn't do what I am expecting, does add_custom_command work in .cmake file

Reply:
I deleted my last comment as I understood it bit later. At present, if a project wants vif.xml file during build, then they can enable it by setting CONFIG_GENERATE_VIF kconfig option to y, no other changes in project. But if I use add_custom_command seems like it needs target and I see existing .cmake files containing add_custom_command are invoked from cmakelists.txt of the project, where I feel kconfig option is simpler for a user enabling it. I am not successful in invoking the add_custom_command without making changes in cmakelists.txt and not successful in find any other project doing it hence assuming it requires changes in project code. I am expecting vif.xml file gets updated when ever there is change in device tree similar to how they reflect in zephyr.dts, hence I am more inclined with existing approach, but if the other approach has other data points I am unaware, I am ready to have a look at them

Reply:
I removed copying to new file and comparing, now the script writes to vif.xml file directly

Reply:
This is not correct, it should only write the file if it needs to be updated. I'll have to look at your issue when I get free time

Reply:
> I kind of picked that mechanism from dts.cmake where zephyr.dts is being generated similar way.

that is because zephyr.dts is needed by the build system itself, hence it must be generated when CMake runs.

The vif is a build output, so that should be generated at build time, and only when input changes, as mentioned by @nordicjm .

CMake provides an example on how to do this here, if the linked example above is not clear enough:
https://cmake.org/cmake/help/book/mastering-cmake/chapter/Custom%20Commands.html#adding-a-custom-target

Reply:
Oh ok, I assumed zephyr.dts is for human readability and used by systems after build, as documentation mentioned it's used for debugging and dtc if enabled. I updated the PR with add_custom_command. Thank you @nordicjm @tejlmand